cmd/dist/fetchobject.go

```diff
@@ -99,91 +86,217 @@ var fetchObjectCommand = cli.Command{
+			return nil, errors.Errorf("unexpected status code %v: %v", u, resp.Status)
+		}
+
+		return resp.Body, nil
+	}
+
+	return nil, errors.New("not found")
+}
+
+func (r *dockerFetcher) doRequest(ctx contextpkg.Context, req *http.Request) (*http.Response, error) {
+	ctx = log.WithLogger(ctx, log.G(ctx).WithField("url", req.URL.String()))
+	log.G(ctx).WithField("request.headers", req.Header).Debug("fetch content")
+	if r.token != "" {
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", r.token))
+	}
+
+	resp, err := ctxhttp.Do(ctx, http.DefaultClient, req)
+	if err != nil {
+		return nil, err
+	}
+	log.G(ctx).WithFields(logrus.Fields{
+		"status":           resp.Status,
+		"response.headers": resp.Header,
+	}).Debug("fetch response received")
+
+	return resp, err
+}
+
+type dockerResolver struct{}
+
+var _ remotes.Resolver = &dockerResolver{}
+
+func newDockerResolver() remotes.Resolver {
+	return &dockerResolver{}
+}
+
+func (r *dockerResolver) Resolve(ctx contextpkg.Context, ref string) (string, ocispec.Descriptor, remotes.Fetcher, error) {
+	refspec, err := reference.Parse(ref)
+	if err != nil {
+		return "", ocispec.Descriptor{}, nil, err
+	}
 
-			paths, err := getV2URLPaths(prefix, object, hints...)
-			if err != nil {
-				return nil, err
+	var (
+		base  url.URL
+		token string
+	)
+
+	switch refspec.Hostname() {
+	case "docker.io":
+		base.Scheme = "https"
+		base.Host = "registry-1.docker.io"
+		prefix := strings.TrimPrefix(refspec.Locator, "docker.io/")
+		base.Path = path.Join("/v2", prefix)
+		token, err = getToken(ctx, "repository:"+prefix+":pull")
+		if err != nil {
+			return "", ocispec.Descriptor{}, nil, err
+		}
+	case "localhost:5000":
+		base.Scheme = "http"
+		base.Host = "localhost:5000"
+		base.Path = path.Join("/v2", strings.TrimPrefix(refspec.Locator, "localhost:5000/"))
+	default:
+		return "", ocispec.Descriptor{}, nil, errors.Errorf("unsupported locator: %q", refspec.Locator)
+	}
+
+	fetcher := &dockerFetcher{
+		base:  base,
+		token: token,
+	}
+
+	var (
+		urls []string
+		dgst = refspec.Digest()
+	)
+
+	if dgst != "" {
+		if err := dgst.Validate(); err != nil {
+			// need to fail here, since we can't actually resolve the invalid
+			// digest.
+			return "", ocispec.Descriptor{}, nil, err
+		}
+
+		// turns out, we have a valid digest, make a url.
+		urls = append(urls, fetcher.url("manifests", dgst.String()))
+	} else {
+		urls = append(urls, fetcher.url("manifests", refspec.Object))
+	}
+
+	// fallback to blobs on not found.
+	urls = append(urls, fetcher.url("blobs", dgst.String()))
+
+	for _, u := range urls {
+		req, err := http.NewRequest(http.MethodHead, u, nil)
+		if err != nil {
+			return "", ocispec.Descriptor{}, nil, err
+		}
+
+		// set headers for all the types we support for resolution.
+		req.Header.Set("Accept", strings.Join([]string{
+			MediaTypeDockerSchema2Manifest,
+			MediaTypeDockerSchema2ManifestList,
+			ocispec.MediaTypeImageManifest,
+			ocispec.MediaTypeImageIndex, "*"}, ", "))
```

Comment:
[Question] Is it compulsory to check type while Resolve/Fetch?
- I understand, It may lead to issue, where we transfer much data, which is not supported, But making this optional could give flexibility.
Checking for supported ``MediaType`` could be done at ``Ingest`` function.
Thinking from perespective of storing images at ``ssh+git`` as registry